seccomp: switch default to ENOSYS

giuseppe opened the pull request:
add the currently blocked syscalls to a deny-list and switch the default to ENOSYS.
    
Signed-off-by: Giuseppe Scrivano <gscrivan@redhat.com>

rhatdan commented:
Tests are not happy.

rhatdan commented on pkg/seccomp/seccomp.json:
```diff
@@ -1,5 +1,6 @@
 {
 	"defaultAction": "SCMP_ACT_ERRNO",
+	"defaultErrnoRet": 38,
```
Should this be a magic number?

rhatdan commented on pkg/seccomp/seccomp.json:
```diff
@@ -399,6 +400,86 @@
+				"vserver"
+			],
+			"action": "SCMP_ACT_ERRNO",
+			"errnoRet": 1
```
Should this be a magic number?

giuseppe replied:
I think it is already too late for errnoRet as we already support it.  Perhaps add a `errnoRetString` that can be used in place of `errnoRet`? 

We could make `defaultErrnoRet` a string, but I don't see a way to convert from a name to an errno list in the std library, so we would need to do the conversion ourselves.

giuseppe commented:
tests are happy now

rhatdan commented:
LGTM
@containers/podman-maintainers PTAL

vrothberg reviewed (commented):
Can you open a test PR against Podman?

We've regressed a number of the times recently in c/common and caught them while vendoring the release into Podman.

giuseppe commented:
> Can you open a test PR against Podman?

good idea.  Opened a PR here: https://github.com/containers/podman/pull/10451

giuseppe commented:
failures don't seem related to the seccomp changes, but are related to the new runc release.  Some tests need to be fixed

rhatdan commented:
@giuseppe Still working on this?

fweimer-rh commented on pkg/seccomp/default_linux.go:
```diff
@@ -44,8 +44,54 @@ func arches() []Architecture {
+		{
+			Names: []string{
+				"bdflush",
+				"clone3",
```
I think this has to be treated the same way as `clone`. Ideally, we would allow this system call because it is required for supporting certain forms of security hardening. But we definitely will need `ENOSYS` for `clone3`.

fweimer-rh commented on pkg/seccomp/default_linux.go:
```diff
@@ -44,8 +44,54 @@ func arches() []Architecture {
+				"io_pgetevents",
+				"io_uring_enter",
+				"io_uring_register",
+				"io_uring_setup",
```
`io_uring_*` should be `ENOSYS`, too.

fweimer-rh commented on pkg/seccomp/default_linux.go:
```diff
@@ -44,8 +44,54 @@ func arches() []Architecture {
+				"pciconfig_write",
+				"pkey_alloc",
+				"pkey_free",
+				"pkey_mprotect",
```
The `pkey_*` functions should really be permitted because they enable security hardening. Otherwise, `ENOSYS`, not `EPERM` is needed for them.

fweimer-rh commented on pkg/seccomp/default_linux.go:
```diff
@@ -44,8 +44,54 @@ func arches() []Architecture {
+				"pkey_alloc",
+				"pkey_free",
+				"pkey_mprotect",
+				"rseq",
```
I do not see any harm in permitting `rseq`. In the future, it will be an important performance enhancement for some workloads.

fweimer-rh commented on pkg/seccomp/default_linux.go:
```diff
@@ -44,8 +44,54 @@ func arches() []Architecture {
+				"io_uring_setup",
+				"kexec_file_load",
+				"kexec_load",
+				"membarrier",
```
`membarrier` should default to `ENOSYS` if it can't be permitted.

fweimer-rh commented on pkg/seccomp/default_linux.go:
```diff
@@ -573,6 +674,19 @@ func DefaultProfile() *Seccomp {
+				"get_mempolicy",
+				"mbind",
+				"set_mempolicy",
+			},
```
Not sure why those are separate from `move_pages` and `migrate_pages` above?

giuseppe replied:
these fails with EPERM only when the container is not granted CAP_SYS_NICE.  Otherwise seccomp won't block them

fweimer-rh replied:
Hmm. `move_pages` and `migrate_pages` seem to have similar constraints. I also believe that these system calls do not require `CAP_SYS_NICE` in all cases. See the discussion of the `MPOL_MF_MOVE_ALL` flag.

giuseppe replied:
do you suggest enabling them in any case?

fweimer-rh replied:
As far as I know, they are similar to `madvise` and not really privileged, so I think they can be enabled.

giuseppe replied:
thanks.  Added a patch to always allow them

giuseppe commented:
@fweimer-rh thanks for the suggestions.  I agree with them and I'll implement them as a follow-up PR.

I am still struggling to get this in as some Podman CI tests are failing with the new ubuntu/fedora images needed for having runc-1.0-rc95.  Progress here: https://github.com/containers/podman/pull/10451

As soon as the tests pass, I will polish the Podman PR to update the images and we can merge this one

giuseppe commented:
Podman tests are finally green.  If we merge the PR, I will cleanup the Podman PR to vendor a release of containers/common and we can then target the new changes to the seccomp policy

rhatdan commented:
/lgtm

giuseppe commented:
@fweimer-rh addressed your comments in https://github.com/containers/common/pull/627, except for mbind, set_mempolicy, get_mempolicy, but I can add another patch to address it
